### PR TITLE
Copter: support MAV_FRAME_LOCAL_NED in MISSION_ITEM for auto mode #8748 (rebased)

### PR DESCRIPTION
**Pull request by roackb2**
This is based on the work of @chobitsfan with branch and issue #8748.
I've done the rebase on master branch and need someone to check the logic is correct.
@amilcarlucas may I kindly request for your review?

### CODE REVIEW

**Review by amilcarlucas — CHANGES_REQUESTED** [review]
remove the temp.cpp file

**Comment by roackb2** [comment]
My bad, removed it.

**Comment by roackb2** [comment]
And now it fails on cloning MAVProxy....
Any idea?

```
Installing MAVProxy
/tmp /__w/ardupilot/ardupilot
+ for t in $CI_BUILD_TARGET
+ '[' sitltest-tracker == sitltest-heli ']'
+ '[' sitltest-tracker == sitltest-copter-tests1 ']'
+ '[' sitltest-tracker == sitltest-copter-tests1a ']'
+ '[' sitltest-tracker == sitltest-copter-tests1b ']'
+ '[' sitltest-tracker == sitltest-copter-tests1c ']'
+ '[' sitltest-tracker == sitltest-copter-tests1d ']'
+ '[' sitltest-tracker == sitltest-copter-tests1e ']'
+ '[' sitltest-tracker == sitltest-copter-tests2 ']'
+ '[' sitltest-tracker == sitltest-copter-tests2a ']'
+ '[' sitltest-tracker == sitltest-copter-tests2b ']'
+ '[' sitltest-tracker == sitltest-can ']'
+ '[' sitltest-tracker == sitltest-plane ']'
+ '[' sitltest-tracker == sitltest-quadplane ']'
+ '[' sitltest-tracker == sitltest-rover ']'
+ '[' sitltest-tracker == sitltest-tracker ']'
+ run_autotest Tracker build.Tracker test.Tracker
+ NAME=Tracker
+ BVEHICLE=build.Tracker
+ RVEHICLE=test.Tracker
+ cat /proc/cpuinfo
+ '[' 0 -eq 0 ']'
+ echo 'Installing MAVProxy'
+ pushd /tmp
+ git clone --recursive https://github.com/ardupilot/MAVProxy
Cloning into 'MAVProxy'...
error: RPC failed; curl 56 GnuTLS recv error (-54): Error in the pull function.
fatal: the remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
```

**Comment by roackb2** [comment]
@amilcarlucas finally it passed all the checks lolllllll
This is my first time doing a PR, but I'm wondering that is an nondeterministic CI/CD pipeline a common phenomenon?

**Comment by amilcarlucas** [comment]
Nice. yes, sometimes the CI is non-deterministic.

**Comment by roackb2** [comment]
@amilcarlucas I've tested this feature with a copter for many times. It is really very helpful for my current project on hand, and I believe it would also be needed for many others studying indoor flight. To support those who might need this, I listed some information as following:

Here's the [video](https://youtu.be/fm_W8DlqdkU)
This is the [flight plan](https://www.dropbox.com/s/xk4rsuaxvxkiv2e/L-line-2.plan?dl=0)
And this is the [parameters](https://www.dropbox.com/s/da6vyfaay2bf9ql/vision_master.params?dl=0) I use

File: ArduCopter/mode_auto.cpp [hunk 1]
```diff
@@ -154,41 +154,52 @@ void ModeAuto::takeoff_start(const Location& dest_loc)
...
+        if (alt_target < 100) {
+            dest.set_alt_cm(100, Location::AltFrame::ABOVE_HOME);
+        }
 
-    // set waypoint controller target
-    if (!wp_nav->set_wp_destination(dest)) {
-        // failure to set destination can only be because of missing terrain data
-        copter.failsafe_terrain_on_event();
-        return;
+	// set waypoint controller target
```
Comment (amilcarlucas) [inline ArduCopter/mode_auto.cpp:197]: The indentation looks messed up

File: libraries/AP_Mission/AP_Mission.cpp [hunk 2]
```diff
@@ -1538,14 +1589,22 @@ bool AP_Mission::mission_cmd_to_mavlink_int(const AP_Mission::Mission_Command& c
...
+            packet.x = cmd.content.location.lat * 100;
+            packet.y = cmd.content.location.lng * 100;
+            packet.z = cmd.content.location.alt * -0.01f;
+            packet.frame = MAV_FRAME_LOCAL_NED;
         } else {
-            packet.frame = MAV_FRAME_GLOBAL;
+            packet.x = cmd.content.location.lat;
+            packet.y = cmd.content.location.lng;
+
+            packet.z = cmd.content.location.alt / 100.0f;   // cmd alt in cm to m
```
Comment (amilcarlucas) [inline libraries/AP_Mission/AP_Mission.cpp:1602]: Please use a multiplication instead of a division here.

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

**Comment by roackb2** [comment]
@amilcarlucas Thanks for your quick response.  I got some errors consistently while running tests multiple times (listed below).

I dig into the code of test cases, looks like the test failed because of the difference in altitude of mission result. The code states that `max_allowed_delta` is 0.000009 for counter 4 ~ 10 in QGC WPL 110 file. The error points out that counter 10, which represents altitude, has difference of 0.000976999999693362 (1085.900391 - 11085.899414), which exceeds the max allowed delta. However, no related error shows up in last commit, and the only change in this PR that relates to altitude is to change the sign, since local frame is NED. The difference in this commit compared to prior one in `AP_Mission.cpp` is to change division to multiplication as you suggested. May there be any reason that this arithmetic change will result in the difference in mission execution? What kind of action should I take?  I'm also curious about that do real vehicle driving or flying could achieve such kind of altitude accuracy?
Thanks.

```
AT-0142.2: Stopping SITL
AT-0143.3: Failing tests:
  DriveMission (Drive Mission balancebot1.txt) (Files have different (float) content: (/vagrant/Tools/autotest/ArduRover_Tests/DriveMission/balancebot1.txt) and (test.Rover/logs/2020-12-06/flight2/way.txt) (2	03	16	0.000000	0.000000	0.000000	0.000000	40.071083	-105.230026	11085.900391	1 vs 2	0	3	16	0.000000	0.000000	0.000000	0.000000	40.071083	-105.230026	11085.899414	1) (11085.900391 vs 11085.899414) (0.0009770000) (count=10)) (see /tmp/buildlogs/Rover-DriveMission.txt)
>>>> FAILED STEP: test.BalanceBot at Sun Dec  6 18:13:27 2020
Failure Summary:
  test.Rover:
    DriveMission (Drive Mission rover1.txt) (Files have different (float) content: (/vagrant/Tools/autotest/ArduRover_Tests/DriveMission/rover1.txt) and (test.Rover/logs/2020-12-06/flight1/way.txt) (2	0	316	0.000000	0.000000	0.000000	0.000000	40.071289	-105.230057	11085.900391	1 vs 2	0	3	16	0.000000	0.000000	0.000000	0.000000	40.071289	-105.230057	11085.899414	1) (11085.900391 vs 11085.899414) (0.0009770000) (count=10)) (see /tmp/buildlogs/Rover-DriveMission.txt)
  test.BalanceBot:
    DriveMission (Drive Mission balancebot1.txt) (Files have different (float) content: (/vagrant/Tools/autotest/ArduRover_Tests/DriveMission/balancebot1.txt) and (test.Rover/logs/2020-12-06/flight2/way.txt) (2	03	16	0.000000	0.000000	0.000000	0.000000	40.071083	-105.230026	11085.900391	1 vs 2	0	3	16	0.000000	0.000000	0.000000	0.000000	40.071083	-105.230026	11085.899414	1) (11085.900391 vs 11085.899414) (0.0009770000) (count=10)) (see /tmp/buildlogs/Rover-DriveMission.txt)
FAILED 2 tests: ['test.Rover', 'test.BalanceBot']
```

**Comment by amilcarlucas** [comment]
I did the fixes and created a new PR #16259

**Comment by amilcarlucas** [comment]
#16259 is cleaner (no temp file), rebased, squashed and passes all CI tests

**Comment by roackb2** [comment]
@amilcarlucas ok thanks for your work

**Pull request closed without merging**